hazelcast/src/main/java/com/hazelcast/internal/metrics/jmx/JmxPublisher.java

```diff
@@ -93,10 +93,16 @@ private void publishNumber(MetricDescriptor originalDescriptor, Number value) {
             return;
         }
 
-        // we need to take a copy of originalDescriptor here to ensure
-        // we map with an instance that doesn't get recycled or mutated
-        MetricDescriptor descriptor = copy(originalDescriptor);
-        MetricData metricData = metricNameToMetricData.computeIfAbsent(descriptor, createMetricDataFunction);
+        final MetricData metricData;
+        if (!metricNameToMetricData.containsKey(originalDescriptor)) {
+            // we need to take a copy of originalDescriptor here to ensure
+            // we map with an instance that doesn't get recycled or mutated
+            MetricDescriptor descriptor = copy(originalDescriptor);
+            metricData = metricNameToMetricData.computeIfAbsent(descriptor, createMetricDataFunction);
+        } else {
+            metricData = metricNameToMetricData.get(originalDescriptor);
```

Comment:
The point of adding the `if` here was to avoid taking a copy of the descriptor (thus to prevent allocation) in the cases when it is already in the map. In the other branch, we take the copy because we will cache it locally and the passed descriptor might be referenced from elsewhere too (from a pool for example). Note that the copy is taken by copying the `originalDescriptor` into a newly created `descriptor` instance and not by calling `originalDescriptor.copy()` where we have no control over where the new instance is from.

On the other hand, it's indeed possible to mutate the descriptor _concurrently_ after passing it to the collector and that case we can get an NPE here and it can probably cause issues elsewhere too. For example, `MetricDescriptorImpl` itself is not thread-safe. It's a trade-off with the mutable descriptors that the API is not safe in terms of concurrency. This mutability of the descriptors has advantages as well that we leverage at some places. Since this is an internal API and we need a concurrently mutating thread - that is at least very smelly -, I think not taking a copy here is fine. Note that once a dynamic provider provides a metric, it gets collected and published immediately on the same thread, before the call returns to the provider.

Reply:
Thanks for the clarification. It makes sense to me.